python/adjoint/objective.py

```diff
@@ -166,16 +166,26 @@ def place_adjoint_source(self,dJ):
 
         self.sources = []
 
+        iomega = (1.0 - np.exp(-1j * (2 * np.pi * self.frequencies) * dt)) * (1.0 / dt) # scaled frequency factor with discrete time derivative fix
+        src = self.time_src
+
+        # an ugly way to calcuate the scaled dtft of the forward source
+        y = np.array([src.swigobj.current(t,dt) for t in np.arange(0,T,dt)]) # time domain signal
+        fwd_dtft = np.matmul(np.exp(1j*2*np.pi*self.frequencies[:,np.newaxis]*np.arange(y.size)*dt), y)*dt/np.sqrt(2*np.pi) # dtft
+
+        # we need to compensate for the phase added by the time envelope at our freq of interest
+        src_center_dtft = np.matmul(np.exp(1j*2*np.pi*np.array([src.frequency])[:,np.newaxis]*np.arange(y.size)*dt), y)*dt/np.sqrt(2*np.pi)
+        adj_src_phase = np.exp(1j*np.angle(src_center_dtft))
```

Comment:
This copy-and-paste code should be refactored into a function.

See e.g. https://www.beyondjava.net/the-art-of-refactoring-removing-code-duplications